### PR TITLE
Generate Unit class (Fixes #2170)

### LINKED ISSUE #2170: Backend: Generate the `Unit` class.

### PR DESCRIPTION
**Pull request by JonathanStarup**
Fixes #2170

### CODE REVIEW

**Comment by JonathanStarup** [comment]
@magnus-madsen two questions: 
Is it correct of me to remove the toString method from the old class? 
How can remove the Unit class from the library files?

**Comment by magnus-madsen** [comment]
> @magnus-madsen two questions:
> Is it correct of me to remove the toString method from the old class?

We do not need any methods. `equals`, `hashCode`, `toString` etc are all handled already (and with a different mechanism).

> How can remove the Unit class from the library files?

I will do that later.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenUnitClass.scala [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+package ca.uwaterloo.flix.language.phase.jvm
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.FinalAst.Root
+import org.objectweb.asm.ClassWriter
+import org.objectweb.asm.Opcodes._
+
+object GenUnitClass {
+
+  val instanceMethodName: String = "getInstance"
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenUnitClass.scala:26]: I think - in our spirit of simplifying things - we should simply have a static field called `INSTANCE` (following Java convention for enums) and not have any getter method at all. That is:

```
final class Unit {
  public static final Unit INSTANCE = new Unit();
  private Unit () {}
}
```

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenUnitClass.scala [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    val methodVisitor = visitor.visitMethod(ACC_PUBLIC + ACC_STATIC, instanceMethodName, AsmOps.getMethodDescriptor(Nil, jvmType), null, null)
+    methodVisitor.visitCode()
+    methodVisitor.visitFieldInsn(GETSTATIC, name.toInternalName, instanceFieldName, jvmType.toDescriptor)
+    methodVisitor.visitInsn(ARETURN)
+    methodVisitor.visitMaxs(999, 999)
+    methodVisitor.visitEnd()
+  }
+
+  def genClassConstructor(jvmType: JvmType, name: JvmName, visitor: ClassWriter): Unit = {
+    val methodVisitor = visitor.visitMethod(ACC_STATIC, "<clinit>", AsmOps.getMethodDescriptor(Nil, JvmType.Void), null, null)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenUnitClass.scala:72]: constructor should be private.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/jvm/GenUnitClass.scala:72]: Woops. I guess it got confused.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenUnitClass.scala (hunk 1 above)
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/jvm/GenUnitClass.scala]: > I think - in our spirit of simplifying things - we should simply have a static field called `INSTANCE` (following Java convention for enums) and not have any getter method at all. That is:
> 
> ...

Fixed now. Ready to merge

**Comment by magnus-madsen** [comment]
Nice!

**Pull request merged**